### PR TITLE
send file descriptors for sparse areas in get region info

### PR DESCRIPTION
**Pull request by tmakatos**
Signed-off-by: Thanos Makatos <thanos.makatos@nutanix.com>

### CODE REVIEW

File: lib/tran_sock.h [hunk 1]
```diff
@@ -125,7 +127,8 @@ vfu_msg(int sock, uint16_t msg_id,
         enum vfio_user_command cmd,
         void *send_data, size_t send_len,
         struct vfio_user_header *hdr,
-        void *recv_data, size_t recv_len);
+        void *recv_data, size_t recv_len,
```
Comment (jlevon) [inline lib/tran_sock.h:130]: vfu_msg() is supposed to be the simpler version: please add a new vfu_msg_fds() that adds the new recv_fds.

File: lib/tran_sock.h [hunk 2]
```diff
@@ -112,9 +113,10 @@ int
 vfu_msg_iovec(int sock, uint16_t msg_id,
```
Comment (jlevon) [inline lib/tran_sock.h:113]: need to update the comment for recv_fds

File: lib/tran_sock.h [hunk 3]
```diff
@@ -89,7 +89,8 @@ vfu_send_error(int sock, uint16_t msg_id,
  */
 int
 vfu_recv(int sock, struct vfio_user_header *hdr, bool is_reply,
```
Comment (jlevon) [inline lib/tran_sock.h:91]: Please add a vfu_recv_fds() so only those who need it have to use the more complicated version.

File: lib/tran_sock.c [hunk 4]
```diff
@@ -214,6 +214,53 @@ vfu_send_error(int sock, uint16_t msg_id,
...
+        *nr_fds = 0;
+    }
+
+    ret = recvmsg(sock_fd, &msg, sock_flags);
+    if (ret == -1) {
+        return -errno;
+    }
+
+    if (msg.msg_flags & MSG_CTRUNC || msg.msg_flags & MSG_TRUNC) {
+        assert(false);
```
Comment (jlevon) [inline lib/tran_sock.c:239]: we really want to do this?

File: include/libvfio-user.h [hunk 5]
```diff
@@ -240,15 +240,16 @@ typedef ssize_t (vfu_region_access_cb_t) (void *pvt, char *buf, size_t count,
  * @flags: region  flags
  * @mmap_areas: array of memory mappable areas
  * @nr_mmap_areas: size of mmap_areas
- * @map: callback function to map region
+ * @fd: file descriptor of the file backing the region if it's a mappable
+ *      region
  *
```
Comment (jlevon) [inline include/libvfio-user.h:245]: Can we get some more docs on how the mmap areas / fd is used? Just something describing that the client will be able to request the given fd, and will mmap the regions based upon the sparse mappings.

Maybe we should also point out that there's nothing stopping the client mapping anything in that fd regardless of the areas passed in. Or maybe that's too obvious.

**Review by jlevon — CHANGES_REQUESTED** [review]

File: include/libvfio-user.h [hunk 6]
```diff
@@ -240,15 +240,16 @@ typedef ssize_t (vfu_region_access_cb_t) (void *pvt, char *buf, size_t count,
  * @flags: region  flags
  * @mmap_areas: array of memory mappable areas
  * @nr_mmap_areas: size of mmap_areas
- * @map: callback function to map region
+ * @fd: file descriptor of the file backing the region if it's a mappable
```
[the remark this answers is not in the document]
Comment (tmakatos) [reply include/libvfio-user.h:243]: right

File: include/libvfio-user.h [hunk 7]
```diff
@@ -240,15 +240,16 @@ typedef ssize_t (vfu_region_access_cb_t) (void *pvt, char *buf, size_t count,
...
+ *      region
  *
  * @returns 0 on success, -1 on error, Sets errno.
  */
 int
 vfu_setup_region(vfu_ctx_t *vfu_ctx, int region_idx, size_t size,
                  vfu_region_access_cb_t *region_access, int flags,
                  struct iovec *mmap_areas, uint32_t nr_mmap_areas,
-                 vfu_map_region_cb_t *map);
+                 int fd);
```
[the remark this answers is not in the document]
Comment (tmakatos) [reply include/libvfio-user.h:252]: right

File: lib/libvfio-user.c [hunk 8]
```diff
@@ -879,7 +897,7 @@ UNIT_TEST_SYMBOL(get_next_command);
 
 int
 exec_command(vfu_ctx_t *vfu_ctx, struct vfio_user_header *hdr, size_t size,
-             int *fds, size_t nr_fds,
+             int *fds, size_t nr_fds, int **fds_out, size_t *nr_fds_out,
```
[the remark this answers is not in the document]
Comment (tmakatos) [reply lib/libvfio-user.c:900]: Yes, I agree. I just want to fix the main issue ASAP to allow the SPDK integration to move on, we should clean it up.

File: lib/libvfio-user.c [hunk 9]
```diff
@@ -1024,8 +1043,8 @@ process_request(vfu_ctx_t *vfu_ctx)
 {
     struct vfio_user_header hdr = { 0, };
     int ret;
-    int *fds = NULL;
-    size_t nr_fds, i;
+    int *fds = NULL, *fds_out = NULL;
+    size_t nr_fds, nr_fds_out = 0, i;
```
[the remark this answers is not in the document]
Comment (tmakatos) [reply lib/libvfio-user.c:1047]: ok

File: lib/tran_sock.h (hunk 1 above)
Comment (tmakatos) [reply lib/tran_sock.h:130]: Good idea.

File: lib/libvfio-user.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (jlevon) [reply lib/libvfio-user.c:900]: Makes sense

File: lib/tran_sock.h (hunk 2 above)
Comment (tmakatos) [reply lib/tran_sock.h:113]: ok

File: lib/tran_sock.h (hunk 3 above)
Comment (tmakatos) [reply lib/tran_sock.h:91]: ok

File: lib/tran_sock.c (hunk 4 above)
Comment (tmakatos) [reply lib/tran_sock.c:239]: No, that's accidental, removing it.

File: include/libvfio-user.h (hunk 5 above)
Comment (tmakatos) [reply include/libvfio-user.h:245]: Sure, will elaborate.

**Comment by tmakatos** [comment]
@jlevon addressed review comments, please take another look.

File: include/libvfio-user.h [hunk 10]
```diff
@@ -236,12 +220,30 @@ typedef ssize_t (vfu_region_access_cb_t) (void *pvt, char *buf, size_t count,
...
- * @flags: region  flags
- * @mmap_areas: array of memory mappable areas
- * @nr_mmap_areas: size of mmap_areas
+ * @region_access: callback function to access region. If the region is memory
+ *  mappable and the client accesses the region or part of sparse area, then
+ *  the callback is not called.
+ * @flags: region flags (VFU_REGION_FLAG_)
+ * @mmap_areas: array of memory mappable areas. This array provides to the
+ *  server greater control of which specific areas should be memory mapped by
+ *  the client. Each element in the @mmap_areas array describes on such area.
```
Comment (jlevon) [inline include/libvfio-user.h:229]: nit, "one such area"

File: include/libvfio-user.h [hunk 11]
```diff
@@ -236,12 +220,30 @@ typedef ssize_t (vfu_region_access_cb_t) (void *pvt, char *buf, size_t count,
...
+ * @nr_mmap_areas: number of sparse areas in @mmap_areas. Must be 0 if the
+ *  region is not memory mappable.
  * @fd: file descriptor of the file backing the region if it's a mappable
- *      region
+ *  region. It is the server's responsibility to create a file suitable for
+ *  memory mapping by the client. Ignored if the region is not memory mappable.
+ *
+ * A note on memory-mappable regions: the client can memory map any part of the
+ * file descriptor, even if not supposed to do so acocrding to @mmap_areas.
+ * There is no way in Linux to avoid this.
```
Comment (jlevon) [inline include/libvfio-user.h:239]: Thanks for the longer explanation, this is great.

**Review by jlevon — APPROVED** [review]
Just a couple of nits and this looks good, thanks.

File: include/libvfio-user.h (hunk 10 above)
Comment (tmakatos) [reply include/libvfio-user.h:229]: ok

**Pull request merged**